src/Reflection/ReflectionMethod.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations\Reflection;
+
+use Doctrine\Annotations\Parser\PhpParser;
+
+/**
+ * Reflection Method
+ *
+ * @author Fabio B. Silva <fabio.bat.silva@gmail.com>
+ */
+class ReflectionMethod extends \ReflectionMethod
+{
+    /**
+     * @var \Doctrine\Annotations\Parser\PhpParser
+     */
+    private $phpParser;
+
+    /**
+     * @var array
+     */
+    private $imports;
+
+    /**
+     * @var \Doctrine\Annotations\Reflection\ReflectionClass
+     */
+    private $declaringClass;
+
+    /**
+     * Constructor.
+     *
+     * @param string                                 $className
+     * @param string                                 $methodName
+     * @param \Doctrine\Annotations\Parser\PhpParser $phpParser
+     */
+    public function __construct(string $className, string $methodName, PhpParser $phpParser)
+    {
+        parent::__construct($className, $methodName);
+
+        $this->phpParser = $phpParser;
+    }
+
+    /**
+     * @return array
+     */
+    public function getImports() : array
+    {
+        if ($this->imports !== null) {
+            return $this->imports;
+        }
+
+        $class        = $this->getDeclaringClass();
+        $classImports = $class->getImports($class);
+        $traitImports = [];
+
+        foreach ($class->getTraits() as $trait) {
+            if ( ! $trait->hasMethod($this->getName())) {
+                continue;
+            }
+
+            if ($trait->getFileName() !== $this->getFileName()) {
+                continue;
+            }
+
+            $methodImports = $this->phpParser->parseClass($trait);
+            $traitImports  = array_merge($traitImports, $methodImports);
+        }
+
+        return $this->imports = array_merge($classImports, $traitImports);
+    }
```

Comment:
I think this is confusing. If the method is coming from a trait, then the annotation is living in the trait file, and the imports should be from the trait's file only. If the method is declared in the class itself, then the annotation is living in the class file, and it should use the imports from the class file. I don't see a case where it should combine imports from different files.

Reply:
Here is how to find out if a method is defined in a trait, https://stackoverflow.com/a/45912866/246724.

Reply:
Btw, imo all of this reflection + inheritance adds unnecessary complexity.

Reply:
Well, forbidding people to use inheritance and traits when using annotations would mean that nobody would migrate to version 2

Reply:
From a functional perspective, it should stay the exact same, maybe with some API changes, but no behavioral ones.

The point here is getting rid of our own hacky parser, using a formalised one (HOA's)

Reply:
Oh, maybe I was unclear.

> all of this reflection + inheritance adds unnecessary complexity.

What I mean is we don't need to inherit from \ReflectionClass to find the imports. Instead, have an ImportFinder or something like that.

Of course, people who write annotated classes should be allowed to use inheritance, and traits!

> From a functional perspective, it should stay the exact same, maybe with some API changes, but no behavioral ones.

If the code in the PR is replicating existing behavior, then it needs to stay this way.
Or, if we agree that the old behavior is wrong, we could have two implementations of ImportFinder or of the class name resolver: One that operates the BC way, another that operates the "correct" way.

Why would we say that "the old behavior is wrong"?

Consider this example:

File T.php:
```php
<?php
namespace Acme\Foo;
use Acme\Annotation\Hello;
trait T {
  /**
   * @Hello("I am an annotation on a trait method.")
   * @Goodbye("I am annotation on a trait method, but the import is in the class file.")
   */
  function foo() {}
}
```

File C.php:
```php
<?php
namespace Acme\Bar;
use Acme\Annotation\Goodbye;
class C {
  use T;
}
```

With the behavior proposed in the PR, which I assume is also the current behavior, the second annotation `@Goodbye(..)` will use the import `Acme\Annotation\Goodbye` from the class file.

I am saying this is wrong. It should only use the imports from the trait file. So the `@Hello(..)` should work, but the `@Goodbye(..)` should not.

This would be consistent with how the language itself works.
Imports are only available within the same file.

Well personally I think having annotations on a method in a trait is probably a bad idea anyway. but if we support it, it should at least be "correct". Unless, of course, it is for BC reasons.

> The point here is getting rid of our own hacky parser, using a formalised one (HOA's)

Which I assume will be more maintainable, more reliable, more understandable (people have to look at the grammar only). So yeah, seems like a good idea.

Personally I care more about the registry going away.

Reply:
About the methods defined in traits: It gets even more interesting if the method is renamed.

```php
Trait T {
  /**
   * @Hello()
   */
  function foo() {}
}

class C {
  use T {
    foo as bar;
  }
}

$m = new \ReflectionMethod('C', 'bar');
$reader->getMethodAnnotations($m);
```

The current behavior will not understand that the method is defined in a trait under a different name.

Reply:
And about properties in traits - this is more difficult. There is no \ReflectionProperty::getFileName().
https://stackoverflow.com/questions/18257158/how-to-extract-start-line-of-a-property-declaration-in-php

As a heuristic, we could say that:
- If none of the traits of the class has a property with the same name, then the property belongs to the class, obviously.
- If one or more of the traits define the property then we compare the doc comment. See https://3v4l.org/KY9nl.

Maybe all of this should be discussed in a separate issue. I only brought it up here because the PR affects the code where this behavior is implemented.

Reply:
@donquixote we kinda fixed all of these horrors in [`roave/better-reflection`](https://github.com/Roave/BetterReflection), although it is not the primary aim here to provide very precise reflection of ugly stuff like traits.